In [26]:
import os
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

In [27]:
project_dir = r"C:\Users\Samar mubark\Desktop\flight_price_prediction"
data_dir = "data"

In [28]:
os.path.join(project_dir, data_dir , f"flight_price.csv")

'C:\\Users\\Samar mubark\\Desktop\\flight_price_prediction\\data\\flight_price.csv'

In [29]:
def get_data(name):
    file_name = f"{name}.csv"
    file_path = os.path.join(project_dir, data_dir , file_name)
    return pd.read_csv(file_path)

In [30]:
flights = get_data("flight_price")
flights

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [31]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [32]:
flights.head()

Airline Date_of_Journey    Source Destination                  Route  \
0       IndiGo      24/03/2019  Banglore   New Delhi              BLR → DEL   
1    Air India       1/05/2019   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways       9/06/2019     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      12/05/2019   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      01/03/2019  Banglore   New Delhi        BLR → NAG → DEL   

  Dep_Time  Arrival_Time Duration Total_Stops Additional_Info  Price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [33]:
flights.dtypes

Airline            object
Date_of_Journey    object
Source             object
Destination        object
Route              object
Dep_Time           object
Arrival_Time       object
Duration           object
Total_Stops        object
Additional_Info    object
Price               int64
dtype: object

In [34]:
flights.duplicated(keep=False).sum()

415

In [35]:
flights.loc[flights.duplicated(keep=False)].sort_values(["Airline","Date_of_Journey","Source","Destination","Route"])

Airline Date_of_Journey    Source Destination                  Route  \
6321  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
9848  Air India      01/03/2019  Banglore   New Delhi  BLR → BOM → AMD → DEL   
572   Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
8168  Air India      03/03/2019  Banglore   New Delhi              BLR → DEL   
1495  Air India       1/04/2019   Kolkata    Banglore  CCU → DEL → COK → BLR   
...         ...             ...       ...         ...                    ...   
2692   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2870   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
3711   SpiceJet      24/03/2019  Banglore   New Delhi              BLR → DEL   
2634    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   
5470    Vistara      24/03/2019  Banglore   New Delhi              BLR → DEL   

     Dep_Time  Arrival_Time Duration Total_Stops  \
6321    08:50  23:55 02 Mar   39h 5m     2 stops   
9848    08:50  23:55 02 Mar   39h 5m     2 stops   
572     21:10         23:55   2h 45m    non-stop   
8168    21:10         23:55   2h 45m    non-stop   
1495    10:00  01:20 02 Apr  15h 20m     2 stops   
...       ...           ...      ...         ...   
2692    05:45         08:35   2h 50m    non-stop   
2870    05:45         08:35   2h 50m    non-stop   
3711    20:30         23:20   2h 50m    non-stop   
2634    11:30         14:10   2h 40m    non-stop   
5470    11:30         14:10   2h 40m    non-stop   

                   Additional_Info  Price  
6321                       No info  17135  
9848                       No info  17135  
572                        No info   7591  
8168                       No info   7591  
1495                       No info  10408  
...                            ...    ...  
2692  No check-in baggage included   4273  
2870  No check-in baggage included   4273  
3711  No check-in baggage included   3873  
2634                       No info   5403  
5470                       No info   5403  

[415 rows x 11 columns]

# upon thorough checking i found out there are 415 duplicate rows which should be removed

## will now create function for cleaning data

In [36]:
def convert_to_minutes(ser):
    return(ser
           .str.split(" " ,  expand =True)
           .set_axis(["hour","minutes"], axis = 1)
           .assign(hour = lambda df_: (
           df_
           .hour
           .str.replace("h", "")
           .astype(int)
           .mul(60)
           ),
            minutes = lambda df_:(
            df_
            .minutes
            .str.replace("m","")
            .fillna("0")
            .astype(int)
            
            )     )
           
           .sum(axis=1)
           )
    

In [37]:
def clean(df):
    return(
    df.drop_duplicates()
        .drop(index=[6474])
        .assign(**{
        col: df[col].str.strip()
        for col in df.select_dtypes(include ="O").columns
    })
    .rename(columns = str.lower)
    .assign(
        airline = lambda df_ : (df_
                               .airline
                               .str.replace("Premium economy","")
                                
                               .str.replace("Business", "")
                               .str.title())
        ,
        date_of_journey = lambda df_ : pd.to_datetime(df_.date_of_journey, dayfirst = True),
        dep_time = lambda  df_ : pd.to_datetime(df_.dep_time).dt.time,
        arrival_time =lambda df_ :pd.to_datetime(df_.arrival_time).dt.time,
        duration = lambda df_ : df_.duration.pipe(convert_to_minutes),
        total_stops = lambda df_ : (
        df_
        .total_stops
        .replace("non-stop" ,"0")
        .str.replace("stops?","", regex = True)
        .pipe(lambda ser: pd.to_numeric(ser))
        
        
        ),additional_info = lambda df_ : df_.additional_info.replace("No info","No Info")
        
    
    
    
    )
    .drop(columns="route")
    
    )

In [38]:
clean_data = clean(flights)

In [39]:
X = clean_data.drop(columns="price")
y = clean_data.price.copy()

In [40]:
X_, X_test, y_, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_, y_, test_size=0.2, random_state=42)

print(X_train.shape, y_train.shape)
print(X_val.shape, y_val.shape)
print(X_test.shape, y_test.shape)

(6695, 9) (6695,)
(1674, 9) (1674,)
(2093, 9) (2093,)


In [41]:

def export_data(X, y, name):
    
    
    file_name = f"{name}.csv"
    file_path = os.path.join(project_dir, data_dir, file_name)

    X.join(y).to_csv(file_path, index=False)

    return pd.read_csv(file_path).head()

In [42]:
export_data(X_train, y_train, "train")

airline date_of_journey    source destination  dep_time arrival_time  \
0  Jet Airways      2019-03-21  Banglore   New Delhi  08:55:00     19:10:00   
1  Jet Airways      2019-03-27     Delhi      Cochin  17:30:00     04:25:00   
2        Goair      2019-03-09  Banglore   New Delhi  11:40:00     14:35:00   
3    Air India      2019-06-12   Kolkata    Banglore  09:25:00     18:30:00   
4  Jet Airways      2019-03-12  Banglore   New Delhi  22:55:00     07:40:00   

   duration  total_stops              additional_info  price  
0       615          1.0  In-flight meal not included   7832  
1       655          1.0  In-flight meal not included   6540  
2       175          0.0                      No Info   7305  
3       545          1.0                      No Info   8366  
4       525          1.0  In-flight meal not included  11087

In [43]:
export_data(X_val, y_val, "val")
export_data(X_test, y_test, "test")

airline date_of_journey    source destination  dep_time arrival_time  \
0  Jet Airways      2019-03-06  Banglore   New Delhi  08:00:00     08:15:00   
1     Spicejet      2019-06-06   Kolkata    Banglore  22:20:00     00:40:00   
2       Indigo      2019-03-18   Kolkata    Banglore  05:30:00     08:20:00   
3       Indigo      2019-06-27   Chennai     Kolkata  19:35:00     21:55:00   
4       Indigo      2019-05-06   Kolkata    Banglore  15:15:00     17:45:00   

   duration  total_stops additional_info  price  
0      1455          1.0         No Info  17996  
1       140          0.0         No Info   3873  
2       170          0.0         No Info   4462  
3       140          0.0         No Info   3597  
4       150          0.0         No Info   4804